shell/platform/darwin/macos/framework/Source/FlutterAppLifecycleDelegate.mm

```diff
@@ -0,0 +1,130 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#import "flutter/shell/platform/darwin/macos/framework/Headers/FlutterAppLifecycleDelegate.h"
+
+#include <AppKit/AppKit.h>
+#include <AppKit/NSApplication.h>
+#include <Foundation/Foundation.h>
+#include <objc/message.h>
+
+#include "flutter/fml/logging.h"
+#include "flutter/fml/paths.h"
+
+@interface FlutterAppLifecycleRegistrar ()
+@end
+
+@implementation FlutterAppLifecycleRegistrar {
+  NSMutableArray* _notificationUnsubscribers;
+
+  // Weak references to registered plugins.
+  NSPointerArray* _delegates;
+}
+
+- (void)addObserverFor:(NSString*)name selector:(SEL)selector {
+  [[NSNotificationCenter defaultCenter] addObserver:self selector:selector name:name object:nil];
+  __block NSObject* blockSelf = self;
+  dispatch_block_t unsubscribe = ^{
+    [[NSNotificationCenter defaultCenter] removeObserver:blockSelf name:name object:nil];
+  };
+  [_notificationUnsubscribers addObject:[unsubscribe copy]];
+}
+
+- (instancetype)init {
+  if (self = [super init]) {
+    _notificationUnsubscribers = [[NSMutableArray alloc] init];
+
+// Using a macro to avoid errors where the notification doesn't match the
+// selector.
+#ifdef OBSERVE_NOTIFICATION
+#error OBSERVE_NOTIFICATION ALREADY DEFINED!
+#else
+#define OBSERVE_NOTIFICATION(SELECTOR) \
+  [self addObserverFor:NSApplication##SELECTOR##Notification selector:@selector(handle##SELECTOR:)]
+#endif
+
+    OBSERVE_NOTIFICATION(WillFinishLaunching);
+    OBSERVE_NOTIFICATION(DidFinishLaunching);
+    OBSERVE_NOTIFICATION(WillBecomeActive);
+    OBSERVE_NOTIFICATION(DidBecomeActive);
+    OBSERVE_NOTIFICATION(WillResignActive);
+    OBSERVE_NOTIFICATION(DidResignActive);
+    OBSERVE_NOTIFICATION(WillTerminate);
+    OBSERVE_NOTIFICATION(WillHide);
+    OBSERVE_NOTIFICATION(DidHide);
+    OBSERVE_NOTIFICATION(WillUnhide);
+    OBSERVE_NOTIFICATION(DidUnhide);
+    OBSERVE_NOTIFICATION(DidChangeScreenParameters);
+    OBSERVE_NOTIFICATION(DidChangeOcclusionState);
+
+#undef OBSERVE_NOTIFICATION
+
+    _delegates = [NSPointerArray weakObjectsPointerArray];
+  }
+  return self;
+}
+
+- (void)dealloc {
+  for (dispatch_block_t unsubscribe in _notificationUnsubscribers) {
+    unsubscribe();
+  }
+  [_notificationUnsubscribers removeAllObjects];
+  _delegates = nil;
+  _notificationUnsubscribers = nil;
+}
+
+static BOOL IsPowerOfTwo(NSUInteger x) {
+  return x != 0 && (x & (x - 1)) == 0;
+}
+
+- (void)addDelegate:(NSObject<FlutterAppLifecycleDelegate>*)delegate {
+  [_delegates addPointer:(__bridge void*)delegate];
+  if (IsPowerOfTwo([_delegates count])) {
+    [_delegates compact];
+  }
```

Comment:
This is what the [iOS engine already uses](https://github.com/flutter/engine/blob/eed12f36f5956a37fb4ba13f1b670a5d41c37edb/shell/platform/darwin/ios/framework/Source/FlutterPluginAppLifeCycleDelegate.mm#L101) for its lifecycle delegate registration, so I was mainly just keeping them similar.

I agree that it's probably overkill, but also probably harmless.  The naive implementation would just not call compact at all, instead of on powers of two, so it's not that different/complex, and potentially saves a little memory.  I'm happy to just nuke it, though, if you think it's adding too much complexity.

Reply:
Thanks for the additional context - agreed it's best to keep them consistent!